File: 0001-two-sum/hash-map/hash-map.cpp
#include <unordered_map>
#include <vector>
using namespace std;

class Solution
{
    public:
        vector<int> twoSum(vector<int>& nums, int target)
        {
            // Create map
            unordered_map<int, int> indicesMap;

            for (int i = 0; i < nums.size(); i++)
            {
                // Calculate the required number needed to add up to the target
                int required = target - nums[i];

                // Check if the required number is in the map
                if (indicesMap.find(required) != indicesMap.end())
                {
                    return {indicesMap[required], i};
                }

                // Store the current number and its index in the map
                indicesMap[nums[i]] = i;
            }

            return {};
        }
};

File: 0001-two-sum/hash-map/hash-map.rs
impl Solution
{
    pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32>
    {
        let mut hashMap = std::collections::HashMap::new();

        for (value, key) in nums.iter().enumerate()
        {
            let required = target - key;

            if let Some(&foundIdx) = hashMap.get(&required)
            {
                return vec![foundIdx as i32, value as i32];
            }
            else
            {
                hashMap.insert(key, value);
            }
        }  

        // No valid solution found
        vec![]
    }
}